packages/cli/src/metadataGeneration/typeResolver.ts

```diff
@@ -152,143 +156,233 @@ export class TypeResolver {
+              };
+            });
+
+          const objectLiteral: Tsoa.NestedObjectLiteralType = {
+            dataType: 'nestedObjectLiteral',
+            properties,
+          };
+          const indexInfos = this.current.typeChecker.getIndexInfosOfType(type);
+          const indexTypes = indexInfos.map(indexInfo => {
+            const typeNode = this.current.typeChecker.typeToTypeNode(indexInfo.type, undefined, ts.NodeBuilderFlags.NoTruncation)!;
+            const type = new TypeResolver(typeNode, this.current, mappedTypeNode, this.context, indexInfo.type).resolve();
+            return type;
+          });
+          if (indexTypes.length) {
+            if (indexTypes.length === 1) {
+              objectLiteral.additionalProperties = indexTypes[0];
+            } else {
+              // { [k: string]: string; } & { [k: number]: number; }
+
+              // A | B is sometimes A type or B type, sometimes optionally accepts both A & B members.
+              // Most people & TSOA thinks that A | B can be only A or only B.
+              // So we can accept this merge
+
+              //Every additional property key assumed as string
+              objectLiteral.additionalProperties = {
+                dataType: 'union',
+                types: indexTypes,
+              };
+            }
+          }
+          return objectLiteral;
+        } else {
+          // Known issues & easy to implement: Partial<string>, Partial<never>, ... But I think a programmer not writes types like this
+          throw new GenerateMetadataError(`Unhandled mapped type has found, flags: ${type.flags}`, this.typeNode);
+        }
       };
-      return objectLiteral;
+
+      const referencer = this.getReferencer();
+      const result: Tsoa.Type = calcMappedType(referencer);
+      return result;
     }
 
-    if (ts.isConditionalTypeNode(this.typeNode) && this.referencer && ts.isTypeReferenceNode(this.referencer)) {
-      const type = this.current.typeChecker.getTypeFromTypeNode(this.referencer);
+    if (ts.isConditionalTypeNode(this.typeNode)) {
+      const referencer = this.getReferencer();
+      const resolvedNode = this.current.typeChecker.typeToTypeNode(referencer, undefined, ts.NodeBuilderFlags.NoTruncation)!;
+      return new TypeResolver(resolvedNode, this.current, this.typeNode, this.context, referencer).resolve();
+    }
 
-      if (type.aliasSymbol) {
-        let declaration = type.aliasSymbol.declarations?.[0] as ts.TypeAliasDeclaration | ts.EnumDeclaration | ts.DeclarationStatement;
-        if (declaration.name) {
-          declaration = this.getModelTypeDeclaration(declaration.name as ts.EntityName) as ts.TypeAliasDeclaration | ts.EnumDeclaration | ts.DeclarationStatement;
+    // keyof
+    if (ts.isTypeOperatorNode(this.typeNode) && this.typeNode.operator === ts.SyntaxKind.KeyOfKeyword) {
+      const type = this.current.typeChecker.getTypeFromTypeNode(this.typeNode);
+      if (type.getFlags() & ts.TypeFlags.Index) {
```

Comment:
Is there a helper that makes this typesafe without casting?

Reply:
Good notice, I have fixed it.